File: db4oj.tests/src/com/db4o/db4ounit/common/querying/CascadeOnDeleteTestCase.java
/* Copyright (C) 2004   Versant Inc.   http://www.db4o.com */

package com.db4o.db4ounit.common.querying;

import com.db4o.config.*;

import db4ounit.*;
import db4ounit.extensions.*;

public class CascadeOnDeleteTestCase extends AbstractDb4oTestCase {
	
	public static class Item {
		public String item;
	}
	
	public static class Holder {
		public Item[] items;
	}
	
	public void testNoAccidentalDeletes() throws Exception {
		assertNoAccidentalDeletes(true, true);
		assertNoAccidentalDeletes(true, false);
		assertNoAccidentalDeletes(false, true);
		assertNoAccidentalDeletes(false, false);
	}
	
	private void assertNoAccidentalDeletes(boolean cascadeOnUpdate, boolean cascadeOnDelete) throws Exception {
		deleteAll(Holder.class);
		deleteAll(Item.class);
		
		ObjectClass oc = fixture().config().objectClass(Holder.class);
		oc.cascadeOnDelete(cascadeOnDelete);
		oc.cascadeOnUpdate(cascadeOnUpdate);
		
		reopen();
		
		Item item = new Item();
		Holder holder = new Holder();
		holder.items = new Item[]{ item };
		db().store(holder);
		db().commit();
		
		holder.items[0].item = "abrakadabra";
		db().store(holder);
		if(! cascadeOnDelete && ! cascadeOnUpdate){
			// the only case, where we don't cascade
			db().store(holder.items[0]);
		}
		
		Assert.areEqual(1, countOccurences(Item.class));
		db().commit();
		Assert.areEqual(1, countOccurences(Item.class));
	}
}

File: db4o.net/Db4objects.Db4o.Tests/Db4objects.Db4o.Tests/Common/Querying/CascadeOnDeleteTestCase.cs
/* Copyright (C) 2004 - 2011  Versant Inc.  http://www.db4o.com */

using Db4oUnit;
using Db4oUnit.Extensions;
using Db4objects.Db4o.Config;
using Db4objects.Db4o.Tests.Common.Querying;

namespace Db4objects.Db4o.Tests.Common.Querying
{
	public class CascadeOnDeleteTestCase : AbstractDb4oTestCase
	{
		public class Item
		{
			public string item;
		}

		public class Holder
		{
			public CascadeOnDeleteTestCase.Item[] items;
		}

		/// <exception cref="System.Exception"></exception>
		public virtual void TestNoAccidentalDeletes()
		{
			AssertNoAccidentalDeletes(true, true);
			AssertNoAccidentalDeletes(true, false);
			AssertNoAccidentalDeletes(false, true);
			AssertNoAccidentalDeletes(false, false);
		}

		/// <exception cref="System.Exception"></exception>
		private void AssertNoAccidentalDeletes(bool cascadeOnUpdate, bool cascadeOnDelete
			)
		{
			DeleteAll(typeof(CascadeOnDeleteTestCase.Holder));
			DeleteAll(typeof(CascadeOnDeleteTestCase.Item));
			IObjectClass oc = Fixture().Config().ObjectClass(typeof(CascadeOnDeleteTestCase.Holder
				));
			oc.CascadeOnDelete(cascadeOnDelete);
			oc.CascadeOnUpdate(cascadeOnUpdate);
			Reopen();
			CascadeOnDeleteTestCase.Item item = new CascadeOnDeleteTestCase.Item();
			CascadeOnDeleteTestCase.Holder holder = new CascadeOnDeleteTestCase.Holder();
			holder.items = new CascadeOnDeleteTestCase.Item[] { item };
			Db().Store(holder);
			Db().Commit();
			holder.items[0].item = "abrakadabra";
			Db().Store(holder);
			if (!cascadeOnDelete && !cascadeOnUpdate)
			{
				// the only case, where we don't cascade
				Db().Store(holder.items[0]);
			}
			Assert.AreEqual(1, CountOccurences(typeof(CascadeOnDeleteTestCase.Item)));
			Db().Commit();
			Assert.AreEqual(1, CountOccurences(typeof(CascadeOnDeleteTestCase.Item)));
		}
	}
}
